impl/src/fmt/display.rs

```diff
@@ -222,10 +222,35 @@ impl<'a> Expansion<'a> {
     /// [`Display::fmt()`]: fmt::Display::fmt()
     fn generate_body(&self) -> syn::Result<TokenStream> {
         match &self.attrs.fmt {
-            Some(fmt) => Ok(quote! { ::core::write!(__derive_more_f, #fmt) }),
+            Some(fmt) => Ok(fmt
+                .delegatable()
+                .then(|| {
```

Comment:
nit: I think this method chain is not very easy to follow for me. I have a feeling that it might benefit readability if it is rewritten using regular match if/else.

Reply:
or maybe have the above match, match on two items: self.attrs.fmt and if the fmt is delagatable. Because I feel like there are 5 distinct cases:
1. normal formatting string
2. formatting string that is delegatable
3. no formatting string on single field variant/struct
4. no formatting string on zero-field variant/struct
5. no formatting string on multi-field variant/struct

I think it would be good to make that clear in the code and probably also add some code comments with those descriptions to each match arm.

Reply:
@JelteF 

> or maybe have the above match, match on two items: `self.attrs.fmt` and if the `fmt` is delegatable.

Just checking whether "`fmt` is delegatable" is not enough here, because it doesn't guarantee that the following `fmt.bindings()` extraction will have exactly one element. The way the code is composed now is that if the `fmt.bindings()` extraction for some reason returns no elements or more than one - we fall back to a normal behavior without delegation to piggyback all the compile-time checks and beautiful errors to the `write!()` macro.

The current code is the most laconic one, without bringing in fancy combinators like `itertools`. I'll think how to make it easier to follow. Since it clearly repeats for `Display` and `Debug`, the most obvious solution is to extract it to another helper method.

Reply:
tl;dr I think we should change `delagatable` to address this (if possible):
> Just checking whether "fmt is delegatable" is not enough here, because it doesn't guarantee that the following fmt.bindings() extraction will have exactly one element. 

I'd expect that these 4 checks would make sure that the bindings extraction would have exactly one element:
```rust
// (1) There is exactly one formatting parameter.
more.is_empty().then_some(p)
// ...
// (3) And either exactly one positional argument is specified.
Some(parsing::Argument::Integer(_)) | None => self.args.len() == 1,
// (4) Or the formatting parameter's name refers to some outer binding.
Some(parsing::Argument::Identifier(_)) if self.args.is_empty() => true,
// (5) Or exactly one named argument is specified for the formatting parameter's name.
Some(parsing::Argument::Identifier(name)) => {
````

Can you give an examlpe of a struct + format string where that isn't the case? And if so maybe we can add another check to `delagatable` to catch that one.